storage/replica.go

```diff
@@ -1150,42 +1150,106 @@ func (r *Replica) beginCmds(ctx context.Context, ba *roachpb.BatchRequest) (func
+			prereq := newCmd.prereqs[i]
 			select {
-			case <-ch:
+			case <-prereq.pending:
+				// Either the prerequisite command finished executing or it was cancelled.
+				// If the command finished cleanly, there's nothing for us to do except wait
+				// for the next prerequisite to finish. However, if the command was cancelled,
+				// we have to handle the cancellation here. This check does not need to be
+				// synchonized, because prereq.cancelled will only ever be set on the other
+				// side of the prereq.pending closing (see below), which the Go Memory Model
+				// promises is safe.
+				if prereq.cancelled {
+					// The first dependent of the cancelled command to get here cleans it up.
+					prereq.cleanUp.Do(func() {
+						r.mu.cmdQ.removeFromTree(prereq)
+					})
+
+					// Migrate transitive dependencies from cancelled prerequisite to
+					// current command. All prerequisites of the prerequisite that was
+					// just cancelled are now this command's direct prerequisites.
+					newCmd.prereqs = append(newCmd.prereqs, prereq.prereqs...)
+				}
 			case <-ctxDone:
 				err := ctx.Err()
 				errStr := fmt.Sprintf("%s while in command queue: %s", err, ba)
 				log.Warning(ctx, errStr)
 				log.ErrEvent(ctx, errStr)
-				go func() {
-					// The command is moot, so we don't need to bother executing.
-					// However, the command queue assumes that commands don't drop
-					// out before their prerequisites, so we still have to wait it
-					// out.
-					for _, ch := range chans {
-						<-ch
+
+				r.mu.Lock()
+				defer r.mu.Unlock()
+				newCmd.cancelled = true
+
+				// Truncate the cancelled commands' prerequisite list to only those
+				// prerequisites that have not yet been waited on.
+				prs := len(newCmd.prereqs)
+				newCmd.prereqs = newCmd.prereqs[i:prs:prs]
+
+				// Each of the cancelled commands' dependents will now have each of the
+				// cancelled commands' prerequisite as direct prerequisites. While locked,
+				// we update all dependents counts on prerequisites to account for this.
+				var fixCommandsPrereqs func(*cmd)
+				fixCommandsPrereqs = func(command *cmd) {
+					for _, prereq := range command.prereqs {
+						prereq.dependents += newCmd.dependents - 1
+
+						// It is possible for prerequisites and dependents to be cancelled at
+						// the same time, so we need to be careful here in case we are a dependent
+						// for an already cancelled prerequisite. This is especially important
+						// if we are the last dependent remaining for the cancelled prereq, because
+						// if we did not handle it here, it would be abandoned.
+						if prereq.cancelled {
+							prereq.cleanUp.DoLocked(func() {
+								r.mu.cmdQ.removeFromTree(prereq)
+							})
+
+							// Our prerequisite cancelled before us and told each of its
+							// prerequisites that we were one of their new dependent commands.
+							// Decrement their dependents count and recursively assure that
```

Comment:
This only decrements if newCmd.dependents is zero. Is that always the case here?